src/ref_.rs

```diff
@@ -0,0 +1,208 @@
+use std::{marker::PhantomData, ops::Deref, ptr::NonNull};
+
+use crate::jvm::JavaObjectExt;
+use crate::thread;
+use crate::{
+    cast::Upcast,
+    jvm::CloneIn,
+    plumbing::ObjectPtr,
+    raw::{EnvPtr, JvmPtr},
+    JavaObject, Jvm,
+};
+
+/// An owned local reference to a non-null Java object of type `T`. The reference will be freed when
+/// dropped. Cannot be shared across threads or [`Jvm::with`] invocations.
+pub struct Local<'jvm, T: JavaObject> {
+    env: EnvPtr<'jvm>,
+    obj: ObjectPtr,
+    _marker: PhantomData<T>,
+}
+
+impl<'jvm, T: JavaObject> Local<'jvm, T> {
+    /// Convert an existing local reference pointed to by `obj` into an owned `Local`. This is used by
+    /// codegen to wrap the output of most JNI calls to prevent the user from not freeing local refs.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `obj` points to a Java object that is an instance of `T` (or its subclasses), is a
+    /// a live, local reference in the current frame, will not later be deleted (including through another call to
+    /// `from_raw()`), and will not dereferenced after the returned [`Local`] is dropped.
+    #[doc(hidden)]
+    pub unsafe fn from_raw(env: EnvPtr<'jvm>, obj: ObjectPtr) -> Self {
+        Self {
+            obj,
+            env,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Creates a *new* local reference to `obj` in the current frame via a `NewLocalRef` JNI call.
+    pub(crate) fn new(env: EnvPtr<'jvm>, obj: &T) -> Self {
+        // SAFETY: The JavaObject trait contract ensures that &T points to a Java object that is an instance of T.
+        unsafe {
+            let new_ref = env.invoke(
+                |jni| jni.NewLocalRef,
+                |jni, f| f(jni, obj.as_raw().as_ptr()),
+            );
+            Self::from_raw(env, NonNull::new(new_ref).unwrap().into())
+        }
+    }
+}
+
+impl<T: JavaObject> Drop for Local<'_, T> {
+    fn drop(&mut self) {
+        // SAFETY: Local owns the local ref and it's no longer possible to dereference the object pointer.
+        unsafe {
+            self.env
+                .invoke(|jni| jni.DeleteLocalRef, |jni, f| f(jni, self.obj.as_ptr()));
+        }
+    }
+}
+
+impl<T: JavaObject> Deref for Local<'_, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: from the guarantees of Local::from_raw, we know obj is a live local ref to an instance of T
+        unsafe { T::from_raw(self.obj) }
+    }
+}
+
+/// An owned global reference to a non-null Java object of type `T`. The reference will be freed when dropped.
+pub struct Global<T: JavaObject> {
+    jvm: JvmPtr,
```

Comment:
given that there is only one jvm per process, do we really need to hold on to this pointer? can't we store it somewhere global?

Reply:
Oops, yes, we definitely can!

Reply:
Removed in next commit